game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java

```diff
@@ -357,46 +357,57 @@ private Territory getSubmerge(final String message) {
         SwingAction.of(
             "Submerge Subs?",
             e -> {
-              final String ok = "Submerge";
-              final String cancel = "Remain";
-              final String wait = "Ask Me Later";
-              final String[] options = {ok, cancel, wait};
-              final int choice =
-                  JOptionPane.showOptionDialog(
-                      BattleDisplay.this,
-                      message,
-                      "Submerge Subs?",
-                      JOptionPane.OK_CANCEL_OPTION,
-                      JOptionPane.PLAIN_MESSAGE,
-                      null,
-                      options,
-                      cancel);
-              // dialog dismissed
-              if (choice == -1) {
-                return;
-              }
-              // wait
-              if (choice == 2) {
-                return;
-              }
-              // remain
-              if (choice == 1) {
+              actionButton.setEnabled(false);
+              if (showSubmergeDialog(message, retreatTo)) {
+                actionButton.setAction(nullAction);
                 latch.countDown();
-                return;
               }
-              // submerge
-              retreatTo.set(battleLocation);
-              latch.countDown();
+              actionButton.setEnabled(true);
             });
-    SwingUtilities.invokeLater(() -> actionButton.setAction(action));
-    SwingUtilities.invokeLater(() -> action.actionPerformed(null));
+    SwingUtilities.invokeLater(
+        () -> {
+          actionButton.setAction(action);
+          action.actionPerformed(null);
+        });
     mapPanel.getUiContext().addShutdownLatch(latch);
     Interruptibles.await(latch);
     mapPanel.getUiContext().removeShutdownLatch(latch);
-    SwingUtilities.invokeLater(() -> actionButton.setAction(nullAction));
     return retreatTo.get();
   }
 
+  private boolean showSubmergeDialog(
+      final String message, final AtomicReference<Territory> retreatTo) {
```

Comment:
Not mandatory, but I think it would make a nicer API if you returned some sort of `RetreatResult` object here:
```java
@AllArgsConstructor
@Getter
class RetreatResult {
  private final boolean confirmed;
  @Nullable
  private final Territory targetTerritory;
}
```
Instead of setting an atomic reference. I'm not quite happy with this approach though, even though it would preserve existing logic.

Instead I'd rather go with this approach:
```java
  @Nullable
  Territory getRetreat(
      final String message, final Collection<Territory> possible, final boolean submerge) {
    final Territory territory = (!submerge || possible.size() > 1)
        ? getRetreatInternal(message, possible)
        : getSubmerge(message);
    return territory.equals(getBattleLocation()) ? null : territory;
}
```

And then both `showSubmergeDialog` and `showRetreatDialog`, return an empty optional at places where false is currently returned and an optional with the retreat territory (the current battle territory instead of null for the remain case) else.
This is a far more "logical" approach and the "current territory" could even get passed out to calling classes to avoid null in a larger refactoring in another PR.

Reply:
I agree that it would be better, but perhaps we can try to minimize such large changes in this PR and do the refactoring without logic changes in a follow-up?

Reply:
No worries. I'll submit a follow-up once this is merged, mostly wrote this for myself so I can understand why it's not that easy. ^^